platform/linux-generic/odp_atomic.c

```diff
@@ -24,3 +26,210 @@ int odp_atomic_lock_free_u64(odp_atomic_op_t *atomic_op)
+		ATOMIC_CAS_OP_128_NO_ORDER(atom, &old, new_val, val);
+
+		if ((val.u64[0] == old.u64[0]) && (val.u64[1] == old.u64[1]))
+			return;
+
+		old = val;
+	}
+}
+
+static int __atomic_cas_u128(odp_atomic_u128_t *atom,
+			     odp_u128_t *old_val, odp_u128_t new_val)
+{
+	int ret = 0;
+	odp_u128_t val;
+
+	ATOMIC_CAS_OP_128_NO_ORDER(atom, old_val, new_val, val);
+
+	if ((val.u64[0] == old_val->u64[0]) && (val.u64[1] == old_val->u64[1]))
+		ret = 1;
+
+	old_val->u64[0] = val.u64[0];
+	old_val->u64[1] = val.u64[1];
+
+	return ret;
+}
+
+static int __atomic_cas_acq_u128(odp_atomic_u128_t *atom,
+				 odp_u128_t *old_val, odp_u128_t new_val)
+{
+	int ret = 0;
+	odp_u128_t val;
+
+	ATOMIC_CAS_OP_128_ACQ(atom, old_val, new_val, val);
+
+	if ((val.u64[0] == old_val->u64[0]) && (val.u64[1] == old_val->u64[1]))
+		ret = 1;
+
+	old_val->u64[0] = val.u64[0];
+	old_val->u64[1] = val.u64[1];
+
+	return ret;
+}
+
+static int __atomic_cas_rel_u128(odp_atomic_u128_t *atom,
+				 odp_u128_t *old_val, odp_u128_t new_val)
+{
+	int ret = 0;
+	odp_u128_t val;
+
+	ATOMIC_CAS_OP_128_REL(atom, old_val, new_val, val);
+
+	if ((val.u64[0] == old_val->u64[0]) && (val.u64[1] == old_val->u64[1]))
+		ret = 1;
+
+	old_val->u64[0] = val.u64[0];
+	old_val->u64[1] = val.u64[1];
+
+	return ret;
+}
+
+static int __atomic_cas_acq_rel_u128(odp_atomic_u128_t *atom,
+				     odp_u128_t *old_val,
+				     odp_u128_t new_val)
+{
+	int ret = 0;
+	odp_u128_t val;
+
+	ATOMIC_CAS_OP_128_ACQ_REL(atom, old_val, new_val, val);
+
+	if ((val.u64[0] == old_val->u64[0]) && (val.u64[1] == old_val->u64[1]))
+		ret = 1;
+
+	old_val->u64[0] = val.u64[0];
+	old_val->u64[1] = val.u64[1];
+
+	return ret;
+}
+
+#else /* Locked version */
+
+static odp_u128_t __atomic_load_u128(odp_atomic_u128_t *atom)
+{
+	return ATOMIC_OP_128(atom, (void)0);
+}
+
+static void __atomic_store_u128(odp_atomic_u128_t *atom, odp_u128_t val)
+{
+	ATOMIC_OP_128(atom, ATOMIC_STORE_OP_128(val));
+}
+
+static int __atomic_cas_u128(odp_atomic_u128_t *atom,
+			     odp_u128_t *old_val, odp_u128_t new_val)
+{
+	int ret;
+	*old_val = ATOMIC_OP_128(atom, ATOMIC_CAS_OP_128(&ret, old_val,
+							 new_val));
+	return ret;
+}
+
+static int __atomic_cas_acq_u128(odp_atomic_u128_t *atom,
+				 odp_u128_t *old_val,
+				 odp_u128_t new_val)
+{
+	int ret;
+	*old_val = ATOMIC_OP_128(atom, ATOMIC_CAS_OP_128(&ret, old_val,
+							 new_val));
+	return ret;
+}
+
+static int __atomic_cas_rel_u128(odp_atomic_u128_t *atom,
+				 odp_u128_t *old_val,
+				 odp_u128_t new_val)
+{
+	int ret;
+	*old_val = ATOMIC_OP_128(atom, ATOMIC_CAS_OP_128(&ret, old_val,
+							 new_val));
+	return ret;
+}
+
+static int __atomic_cas_acq_rel_u128(odp_atomic_u128_t *atom,
+				     odp_u128_t *old_val,
+				     odp_u128_t new_val)
+{
+	int ret;
+	*old_val = ATOMIC_OP_128(atom, ATOMIC_CAS_OP_128(&ret, old_val,
+							 new_val));
+	return ret;
+}
+
+#endif
+
+void odp_atomic_init_u128(odp_atomic_u128_t *atom, odp_u128_t val)
+{
+	__atomic_store_u128(atom, val);
```

Comment:
Petri noticed that in lock-based mode the lock should be initialized before first usage.